### ISSUE TITLE
Crash if element filename contains '@'

### ISSUE THREAD

**Issue by cs-shadow** [cs-shadow/buildstream#70]
In GitLab by @samthursfield on Aug 17, 2017, 10:34

BuildStream constructs the element name from its filename, and in turn constructs an OSTree ref name from the element name.

The Baserock definitions contain some chunks with '@' in their name. It seems '@' is invalid in OSTree ref names, leading to a cras when checking the cache:

```
  File "/usr/lib/python3.5/site-packages/buildstream/element.py", line 719, in _cached
    self.__cached = self.__artifacts.contains(self)
  File "/usr/lib/python3.5/site-packages/buildstream/_artifactcache/artifactcache.py", line 100, in contains
    return _ostree.exists(self.repo, ref)
  File "/usr/lib/python3.5/site-packages/buildstream/_ostree.py", line 203, in exists
    ref = checksum(repo, ref)
  File "/usr/lib/python3.5/site-packages/buildstream/_ostree.py", line 255, in checksum
    _, checksum = repo.resolve_rev(ref, True)
GLib.Error: g-io-error-quark: Invalid refspec baserock/genivi-genivi-demo-platform-audiomanager@gdp/2d8556592e6eee107a1d36ee287f1c31260d6b8b476af427c96e416c90ae998c (0)
```

Full log at: https://gitlab.com/baserock/definitions/-/jobs/29346832

**Comment by cs-shadow** [comment]
In GitLab by @juergbi on Aug 17, 2017, 10:57

This should be handled in artifactcache.buildref(). We may want to switch to a proper escaping mechanism to avoid  name clashes.

**Comment by cs-shadow** [comment]
In GitLab by @samthursfield on Aug 17, 2017, 10:58

Looks like we already escape '+' in ref names: https://gitlab.com/BuildStream/buildstream/blob/master/buildstream/_artifactcache/artifactcache.py#L40

BuildStream appears to convert the filename directly into the element name, meaning that in theory the element name can contain any kind of binary garbage.

OStree is rather more strict:

```
#define OSTREE_REF_FRAGMENT_REGEXP "[-._\\w\\d]+"
#define OSTREE_REF_REGEXP "(?:" OSTREE_REF_FRAGMENT_REGEXP "/)*" OSTREE_REF_FRAGMENT_REGEXP
#define OSTREE_REMOTE_NAME_REGEXP OSTREE_REF_FRAGMENT_REGEXP
```

(from <https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-core.c#L137>)

So we need to escape anything that isn't `-`, `.`, `_` or alphanumeric when generating OSTree refs.

**Comment by cs-shadow** [comment]
In GitLab by @samthursfield on Aug 17, 2017, 11:04

mentioned in commit baserock/definitions@ed666d68c72c9b1f2c97d81a07ab216c3ac1a35c

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 21, 2017, 20:55

closed via commit a10b9ce7dab55490afc85b130852a64633863961

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 21, 2017, 20:55

@samthursfield thanks for finding the accurate information !

Fixed this with the following commit:

```
commit a10b9ce7dab55490afc85b130852a64633863961
Author: Tristan Van Berkom <tristan.vanberkom@codethink.co.uk>
Date:   Mon Aug 21 16:53:50 2017 -0400

    artifactcache.py: Fix escaping of element names
    
    Fixes issue #66
```

**Comment by cs-shadow** [comment]
In GitLab by @juergbi on Aug 22, 2017, 06:41

mentioned in commit a10b9ce7dab55490afc85b130852a64633863961
